File: problems-adapt/0322_fewest-coins-to-make-change/solution_dp.py
from typing import List, Optional


class Solution:
    def fewestCoins(self, coins: List[int], amount: int) -> int:
        # dp[a] = fewest coins for amount a; dp[0] = 0, every other amount
        # starts unreachable (inf doubles as the no-solution marker).
        inf = float("inf")
        dp = [0] + [inf] * amount
        # Amounts smallest-first, so dp[a - c] is already final when consulted.
        for a in range(1, amount + 1):
            best = inf
            # Try every coin as the last one used: dp[a] = min(dp[a - c] + 1).
            for c in coins:
                if c <= a and dp[a - c] + 1 < best:
                    best = dp[a - c] + 1
            dp[a] = best
        return -1 if dp[amount] == inf else dp[amount]

File: problems/0301-0400/0322_fewest-coins-to-make-change/solution_dp.cpp
class Solution {
  public:
    int fewestCoins(vector<int> &coins, int amount) {
        // dp[a] = fewest coins for amount a; dp[0] = 0, every other amount
        // starts unreachable (INF doubles as the no-solution marker).
        const int INF = numeric_limits<int>::max();
        vector<int> dp(amount + 1, INF);
        dp[0] = 0;
        // Amounts smallest-first, so dp[a - c] is already final when consulted.
        for (int a = 1; a <= amount; ++a) {
            // Try every coin as the last one used: dp[a] = min(dp[a - c] + 1).
            for (int c : coins) {
                if (c <= a && dp[a - c] != INF && dp[a - c] + 1 < dp[a]) {
                    dp[a] = dp[a - c] + 1;
                }
            }
        }
        return dp[amount] == INF ? -1 : dp[amount];
    }
};
